src/vs/workbench/contrib/notebook/browser/notebookExecutionStateServiceImpl.ts

```diff
@@ -140,51 +140,68 @@ export class NotebookExecutionStateService extends Disposable implements INotebo
 			exe.onDidComplete(lastRunSuccess => this._onCellExecutionDidComplete(notebookUri, cellHandle, exe, lastRunSuccess)));
 		this._cellListeners.set(CellUri.generate(notebookUri, cellHandle), disposable);
 
-		notebook.onWillAddRemoveCells((e: NotebookTextModelWillAddRemoveEvent) => {
-			const lastFailedCell = this.getLastFailedCellForNotebook(notebook.uri);
-			e.rawEvent.changes.forEach(([start, deleteCount]) => {
-				if (deleteCount) {
-					const deletedHandles = notebook.cells.slice(start, start + deleteCount).map(c => c.handle);
-					if (lastFailedCell && deletedHandles.includes(lastFailedCell)) {
-						this._setLastFailedCellVisibility(notebook.uri, false);
-					}
-				}
-			});
-		});
+		return exe;
+	}
 
-		notebook.onDidChangeContent((events) => {
-			const lastFailedCell = this.getLastFailedCellForNotebook(notebook.uri);
-			events.rawEvents.forEach((e) => {
-				if (e.kind === NotebookCellsChangeType.ModelChange) {
-					e.changes.forEach(([start, deleteCount]) => {
-						if (deleteCount === 0) {
-							const addHandle = notebook.cells[start].handle;
-							if (addHandle === lastFailedCell) {
-								this._setLastFailedCellVisibility(notebook.uri, true);
-							}
-						}
-					});
-				}
-			});
-		});
+	private _setLastFailedCell(notebookURI: URI, cellHandle: number) {
+		this._lastFailedCells.set(notebookURI, cellHandle);
 
-		return exe;
+		if (this._lastFailedCellDisposableListeners.get(notebookURI) === undefined) {
```

Comment:
I think it would be better to do this check inside `_setFailedCellListeners` instead. That just makes it more obvious when I look at that disposable, that it will be tracked correctly and not lost.